src/core/coap/coap_message.cpp

```diff
@@ -322,6 +382,137 @@ otError Message::GetOptionValue(void *aValue) const
     return error;
 }
 
+#if OPENTHREAD_CONFIG_COAP_BLOCKWISE_TRANSFER_ENABLE
+otError Message::ReadBlockOptionValues(uint16_t aOptionLength)
+{
+    otError error                     = OT_ERROR_NONE;
+    uint8_t buf[kMaxOptionHeaderSize] = {0};
+
+    SuccessOrExit(error = GetOptionValue(buf));
+
+    SetBlockWiseBlockNumber(0);
+    SetMoreBlocksFlag(false);
+
+    switch (aOptionLength)
+    {
+    case 1:
+        SetBlockWiseBlockNumber((buf[0] & 0xf0) >> 4);
+        if ((buf[0] & 0x08) >> 3 == 1)
+        {
+            SetMoreBlocksFlag(true);
+        }
+        SetBlockWiseBlockSize((otCoapOptionBlockSize)(buf[0] & 0x07));
+        break;
+    case 2:
+        SetBlockWiseBlockNumber((buf[0] << 4) + ((buf[1] & 0xf0) >> 4));
+        if ((buf[1] & 0x08) >> 3 == 1)
+        {
+            SetMoreBlocksFlag(true);
+        }
+        SetBlockWiseBlockSize((otCoapOptionBlockSize)(buf[1] & 0x07));
+        break;
+    case 3:
+        SetBlockWiseBlockNumber((buf[0] << 12) + (buf[1] << 4) + ((buf[2] & 0xf0) >> 4));
+        if ((buf[2] & 0x08) >> 3 == 1)
+        {
+            SetMoreBlocksFlag(true);
+        }
+        SetBlockWiseBlockSize((otCoapOptionBlockSize)(buf[2] & 0x07));
+        break;
+    default:
+        error = OT_ERROR_INVALID_ARGS;
+        break;
+    }
+
+exit:
+    return error;
+}
+#endif
+
+otError Message::ReadBlockOptionValues(uint16_t               aOptionLength,
+                                       uint32_t *             aBlockNumber,
+                                       bool *                 aMoreBlocks,
+                                       otCoapOptionBlockSize *aBlockSize)
+{
+    otError error                     = OT_ERROR_NONE;
+    uint8_t buf[kMaxOptionHeaderSize] = {0};
+
+    SuccessOrExit(error = GetOptionValue(buf));
+
+    *aBlockNumber = 0;
+    *aMoreBlocks  = false;
+
+    switch (aOptionLength)
```

Comment:
This would cover most use cases, but it seems this is a repetitive way to go about it.

This routine could benefit greatly from https://github.com/openthread/openthread/pull/4396/commits/f17e2edb663cac04b8c61cee8dbf27fae04e3224.

Reply:
That would have also saved some lines of code, unfortunatly, at the time of writing, this feature has already been removed again.